### PR TITLE
Provide CONCAT as a Zephyr API instead of in ARM arch

### LINKED ISSUE #64841: include/zephyr/arch/arm/irq.h defines CONCAT and TOSTR globally, missleading users
**Describe the bug**
[include/zephyr/arch/arm/irq.h](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/arch/arm/irq.h#L79-L84) defines the macros CONCAT and TOSTR which are fully equivalent to Zephyr's [_CONCAT and STRINGIFY macros](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/toolchain/common.h#L133-L138).

This can easily confuse developers who will use these macros assuming they are provided by the OS, and result in code which is not portable between architectures.

**To Reproduce**
N/A

**Expected behavior**
Internal architecture headers not defining macros that can easily be misunderstood as provided by the OS.

**Impact**
Miss-use of these macros leading to code which is not portable between architectures.

**Logs and console output**
N/A

**Environment (please complete the following information):**
N/A

**Additional context**
Cleaning all uses of these macros in tree in https://github.com/zephyrproject-rtos/zephyr/pull/64839

### PR DESCRIPTION
**Pull request by aescolar**
Right now [include/zephyr/arch/arm/irq.h](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/arch/arm/irq.h#L79-L84) defines the macros CONCAT and TOSTR which are fully equivalent to Zephyr's [_CONCAT and STRINGIFY macros](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/toolchain/common.h#L133-L138).

This can easily confuse developers who will use these macros assuming they are provided by the OS, and result in code which is not portable between architectures.

Instead let's remove this definitions from the architecture headers, and remove the only place where TOSTR was used.
Note that all uses of CONCAT in the tree were changed to _CONCAT in https://github.com/zephyrproject-rtos/zephyr/pull/64839

Let's also provide CONCAT as a Zephyr API. (this commit can be dropped if it is too controversial). This should minimize possible issues for users due to the removal from the arm architecture.

Note this PR does not deprecate _CONCAT, but instead leaves it as an internal Zephyr macro.

----

    arm arch: Replace use of TOSTR with STRINGIFY
    
    One of the ARM architure files, defined since long ago
    TOSTR having the exact same purpose as Zephyr's STRINGIFY.
    
    Remove the use of this macro in the tree
    (only used in another ARM header file) in favour
    of STRINGIFY.

-----

    arm arch: Remove definition of CONCAT and TOSTR
    
    One of the ARM architure files, defined since long ago
    CONCAT and TOSTR having the exact same purpose as Zephyr's
    _CONCAT & STRINGIFY.
    
    This arm header file is included thru dependencies into
    almost all code built for ARM, which leads to these
    macros being usable everywhere.
    
    This can easily make developers belive the macros
    are provided by Zephyr itself, and use them, leading
    to code which is not portable between architectures.
    
    Remove this macros definitions from the architecture
    headers.

-----

    include: util: Add CONCAT as a general Zephyr utility macro
    
    Provide the CONCAT macro as a general Zephyr utility macro
    to paste two tokens during the preprocessor pass.
    
    Note that this macro is based on the _CONCAT macro defined
    in toolchain/common.h. This toolchain header needs a CONCAT
    like macro, but requires minimal include dependencies.
    So we leave _CONCAT where it is.

----

Fixes: #64841

### CODE REVIEW

**Review by nordicjm — DISMISSED** [review]

File: include/zephyr/sys/util.h [hunk 1]
```diff
@@ -250,6 +250,17 @@ extern "C" {
 		((type *)(((char *)(ptr)) - offsetof(type, field))); \
 	})
 
+#ifndef CONCAT
```
Comment (gmarull) [inline include/zephyr/sys/util.h:253]: #ifndef needs to be removed, see https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-3-macro-name-collisions

We still have some using ifndef that predate such rule.

**Review by gmarull — CHANGES_REQUESTED** [review]

**Comment by microbuilder** [comment]
@aescolar Failures in CI looks unrelated. Do you mind rebasing, though?

**Comment by aescolar** [comment]
> @aescolar Failures in CI looks unrelated. Do you mind rebasing, though?

The fix seems pending still https://github.com/zephyrproject-rtos/zephyr/pull/65217

**Review by gmarull — APPROVED** [review]

**Review by ycsin — APPROVED** [review]
needs a rebase

**Review by microbuilder — APPROVED** [review]

**Review by ithinuel — APPROVED** [review]

**Review by povergoing — APPROVED** [review]

**Pull request merged**